runtime/src/accounts.rs

```diff
@@ -338,44 +338,46 @@ impl Accounts {
-                        })? {
-                            SystemAccountKind::System => 0,
-                            SystemAccountKind::Nonce => {
-                                // Should we ever allow a fees charge to zero a nonce account's
-                                // balance. The state MUST be set to uninitialized in that case
-                                rent_collector.rent.minimum_balance(nonce::State::size())
-                            }
-                        };
-
-                    if payer_account.lamports() < fee + min_balance {
-                        error_counters.insufficient_funds += 1;
-                        Err(TransactionError::InsufficientFundsForFee)
-                    } else {
-                        payer_account
-                            .checked_sub_lamports(fee)
-                            .map_err(|_| TransactionError::InsufficientFundsForFee)?;
-
-                        let message = tx.message();
-                        let loaders = message
-                            .program_instructions_iter()
-                            .map(|(program_id, _ix)| {
-                                self.load_executable_accounts(ancestors, program_id, error_counters)
-                            })
-                            .collect::<Result<TransactionLoaders>>()?;
-                        Ok(LoadedTransaction {
-                            accounts,
-                            loaders,
-                            rent: tx_rent,
-                            rent_debits,
-                        })
+                    return Err(TransactionError::AccountNotFound);
+                }
+                let min_balance = match get_system_account_kind(payer_account).ok_or_else(|| {
+                    error_counters.invalid_account_for_fee += 1;
+                    TransactionError::InvalidAccountForFee
+                })? {
+                    SystemAccountKind::System => 0,
+                    SystemAccountKind::Nonce => {
+                        // Should we ever allow a fees charge to zero a nonce account's
+                        // balance. The state MUST be set to uninitialized in that case
+                        rent_collector.rent.minimum_balance(nonce::State::size())
                     }
+                };
+
+                if payer_account.lamports() < fee + min_balance {
+                    error_counters.insufficient_funds += 1;
+                    return Err(TransactionError::InsufficientFundsForFee);
                 }
+                payer_account
+                    .checked_sub_lamports(fee)
+                    .map_err(|_| TransactionError::InsufficientFundsForFee)?;
+
+                let program_indices = message
+                    .instructions()
+                    .iter()
+                    .map(|instruction| {
+                        self.load_executable_accounts(
+                            ancestors,
+                            &mut accounts,
```

Comment:
Why do we need to add executable accounts to `accounts` here? Couldn't we add them above?

Reply:
That is a preparation for the next step:
Currently we leave gaps for the executable accounts (so that the indices stay correct) and then load them for each instruction. Thus, they are loaded duplicate if multiple instructions use the same executable. And that could be avoided here, once we can make sure that accounts used as programs are not writable. (see #19629)

Reply:
Ah yeah that is tricky. Someone could update program and then call it later within the same transaction, thanks for the explanation